### PR TITLE
[ELY-1342] Exception for empty mechanisms configuration and avoid …

### PR DESCRIPTION
**Pull request by soul2zimate**
…to try uselessly all mechanisms.

(cherry picked from commit cf51f7060574fc5d9208731bab58f2ed8c96b3be)

master branch PR. JBEAP issue: https://issues.jboss.org/browse/JBEAP-12072
I created https://issues.jboss.org/browse/WFCORE-3100, but I think this should be fixed in Elytron. If this is accepted, I'll move it to an Elytron issue.

### CODE REVIEW

**Comment by darranl** [comment]
Overall I think it is too late in the process to consider a change like this, the 1.1.x branch is on the verge of being tagged Final so making a change like this to the core of Elytron would be a potential compatibility issue.

Overall I think we already have a workaround in the subsystem where a match-all configuration can be defined if we really want all of the mechanisms to be usable, also we may want to add a flag to return a default config in the event of no match.

Would also say this change would mean going from a situation where we have a clear error message stating the no match scenario to a more generic error message only occurring if a mechanism requests a server credential.

**Comment by soul2zimate** [comment]
But this one is for master branch which is targeting 1.2.0.Beta1-SNAPSHOT.

Should I close this one due to the existing workaround and result of different error messages?

**Comment by soul2zimate** [comment]
I have moved WFCORE-3100 to ELY-1342 and update accordingly in commit message and PR title.

File: src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java [hunk 1]
```diff
@@ -967,7 +967,11 @@ private void handleOne(final Callback[] callbacks, final int idx) throws IOExcep
                 } else if (callback instanceof ServerCredentialCallback) {
                     final ServerCredentialCallback serverCredentialCallback = (ServerCredentialCallback) callback;
 
-                    CredentialSource serverCredentialSource = stateRef.get().getMechanismConfiguration().getServerCredentialSource();
+                    MechanismConfiguration mechnismConfiguration  = stateRef.get().getMechanismConfiguration();
+                    if (mechnismConfiguration == null) {
+                        throw log.emptyMechnismConfiguration();
+                    }
+                    CredentialSource serverCredentialSource = mechnismConfiguration.getServerCredentialSource();
```
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: please fix this type, which you repeat on most of lines of your commit: "mechnism" -> "mechanism"

**Comment by darranl** [comment]
Just placing this one on hold for now as I would still like to consider the consequences of the changes further.

**Comment by fjuma** [comment]
Closing this one since it changes existing behaviour.

**Pull request closed without merging**